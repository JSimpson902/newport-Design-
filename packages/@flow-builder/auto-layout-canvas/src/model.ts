import NodeType from './NodeType';

export enum GOTO_CONNECTION_SUFFIX {
    DEFAULT = 'default',
    FAULT = 'fault',
    IMMEDIATE = 'immediate'
}

export type Guid = string;
export type GoToSourceRef = string;

export interface FlowModel {
    [key: string]: NodeModel | ParentNodeModel | BranchHeadNodeModel | (ParentNodeModel & BranchHeadNodeModel);
}
export type NodeRef = Guid | null;
export interface HighlightInfo {
    highlightNext?: boolean;
    highlightLoopBack?: boolean;
    branchIndexesToHighlight?: Array<number>;
    mergeBranchIndexesToHighlight?: Array<number>;
}

export interface ConnectionSource {
    guid: Guid;
    childIndex?: number | null;
}

export interface NodeModel {
    guid: Guid;
    label: string;
    elementType: string;
    maxConnections: number;
    isCanvasElement: boolean;
    config: {
        isSelected: boolean;
        isHighlighted: boolean;
        isSelectable: boolean;
        hasError: boolean;
        highlightInfo?: HighlightInfo | null;
    };

    defaultConnectorLabel?: string;

    // The auto-layout node type corresponding to the element
    nodeType: NodeType;

    // connections
    prev: NodeRef;
    next: NodeRef;
    fault: NodeRef;

    // goto
    incomingGoTo?: GoToSourceRef[];
}

export interface StartNodeModel extends NodeModel {
    triggerType?: string;
    recordTriggerType?: string;
    object?: string;
    startDate?: string;
    startTime?: string;
    filters?: string;
    frequency?: string;
    filterLogic?: string;
    children?: NodeRef[];
    doesRequireRecordChangedToMeetCriteria?: boolean;
}

export interface ParentNodeModel extends NodeModel {
    children: NodeRef[];
    childReferences: Array<{ childReference: string }>;
}

export interface BranchHeadNodeModel extends NodeModel {
    parent: Guid;
    childIndex: number;
    isTerminal: boolean;
}

export interface ElementsMetadata {
    [key: string]: ElementMetadata;
}

export interface MenuItem {
    guid: string;
    icon: string;
    iconContainerClass: string;
    iconClass: string;
    iconSize: string;
    iconVariant: string;
    label: string;
    elementType: string;
    rowClass: string;
    value?: string;
    dataType?: string;
}

export interface MenuSection {
    guid: string;
    heading: string;
    items: MenuItem[];
    label: string;
    separator: boolean;
}

export interface ElementMetadata {
    type: NodeType;
    icon: string;
    section: string;
    description: string;
    iconShape: string;
    iconSize?: string;
    iconBackgroundColor: string;
    label: string;
    elementType: string;
    elementSubtype?: string;
    isSupported: boolean;
    dynamicNodeComponent?: string;
    dynamicNodeComponentSelector?: Function;
}

export const FAULT_INDEX = -1;
export const LOOP_BACK_INDEX = 0;
export const START_IMMEDIATE_INDEX = 0;

const ELEMENT_METADATA_DEFAULT = {
    type: NodeType.DEFAULT,
    icon: 'standard:default'
};

/**
 * @param type - The node type
 * @returns - True if the node can have children
 */
function canHaveChildren(type: NodeType): boolean {
    return type === NodeType.LOOP || type === NodeType.BRANCH || type === NodeType.ROOT;
}

/**
 * @param flow - The flow model
 * @returns - The root node
 */
function getRootNode(flow: FlowModel): ParentNodeModel {
    return flow.root as ParentNodeModel;
}

/**
 * @param flow - The flow model
 * @param key - The guid node
 * @returns The element
 */
function resolveNode(flow: FlowModel, key: NodeRef): NodeModel {
    if (!key) {
        throw new Error('got null guid');
    }

    return flow[key] as NodeModel;
}

/**
 * @param elementsMetadata - Contains elementType -> data map
 * @param elementType - The current node's NodeType
 * @returns - The element metadata
 */
function getElementMetadata(elementsMetadata: ElementsMetadata, elementType: string): ElementMetadata {
    return elementsMetadata[elementType] || ELEMENT_METADATA_DEFAULT;
}

export { resolveNode, getRootNode, canHaveChildren, getElementMetadata };
