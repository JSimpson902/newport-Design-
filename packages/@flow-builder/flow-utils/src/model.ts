import ElementType from './ElementType';

export type guid = string;

export interface FlowModel {
    [key: string]: NodeModel;
}
export type NodeRef = guid | null;

export interface NodeModel {
    guid: guid;
    label: string;
    elementType: string;
    fault: NodeRef;
    maxConnections: number;
    config: { isSelected: boolean; isHighlighted: boolean; canSelect: boolean };

    outcomeReferences?: Array<{ outcomeReference: string }>;
    waitEventReferences?: Array<{ waitEventReference: string }>;
    defaultConnectorLabel?: string;

    // link properties
    prev: NodeRef;
    next: NodeRef;
}

export interface ParentNodeModel extends NodeModel {
    children: NodeRef[];
}

export interface BranchHeadNodeModel extends NodeModel {
    parent: NodeRef;
    childIndex: number;
    isTerminal: boolean;
}

export interface ElementsMetadata {
    [key: string]: ElementMetadata;
}
export interface ElementMetadata {
    type: ElementType;
    icon: string;
}

const ELEMENT_METADATA_DEFAULT = {
    type: ElementType.DEFAULT,
    icon: 'standard:default'
};

function canHaveChildren(type: ElementType): boolean {
    return type === ElementType.LOOP || type === ElementType.BRANCH || type === ElementType.ROOT;
}

function getRootNode(flow: FlowModel): NodeModel {
    return flow.root;
}

function resolveNode(flow: FlowModel, key: NodeRef): NodeModel {
    if (!key) {
        throw new Error('got null guid');
    }

    return flow[key] as NodeModel;
}

function getElementMetadata(elementsMetadata: ElementsMetadata, elementType: string): ElementMetadata {
    return elementsMetadata[elementType] || ELEMENT_METADATA_DEFAULT;
}

export { resolveNode, getRootNode, canHaveChildren, getElementMetadata };
