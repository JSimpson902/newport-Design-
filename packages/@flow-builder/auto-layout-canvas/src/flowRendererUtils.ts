import ConnectorType from './ConnectorTypeEnum';
import ConnectorLabelType from './ConnectorLabelTypeEnum';
import { SvgInfo, Geometry } from './svgUtils';
import {
    FlowModel,
    ElementsMetadata,
    ElementMetadata,
    NodeRef,
    NodeModel,
    BranchHeadNodeModel,
    ParentNodeModel,
    Guid
} from './model';
import MenuType from './MenuType';

export interface LayoutInfo {
    x: number;
    y: number;
    w: number;
    h: number;
    joinOffsetY: number;
    offsetX: number;
    addOffset: number;
    labelOffset: number;
}

export interface NodeLayout {
    prevLayout?: LayoutInfo | undefined;
    layout: LayoutInfo;
}

export interface NodeLayoutMap {
    [key: string]: NodeLayout;
}

export type NodeDimensionMap = Map<Guid, Dimension>;

export interface FlowRenderContext {
    flowModel: FlowModel;
    nodeLayoutMap: NodeLayoutMap;
    progress: number;
    interactionState: FlowInteractionState;
    elementsMetadata: ElementsMetadata;
    layoutConfig: LayoutConfig;
    isFault: boolean;
    isDeletingBranch: boolean;
    dynamicNodeDimensionMap: NodeDimensionMap;
}

export interface InteractionMenuInfo {
    key: Guid;
    type: MenuType;
    geometry?: Geometry;
    needToPosition: boolean;
}

export interface FlowInteractionState {
    menuInfo: InteractionMenuInfo | null;
    deletionPathInfo: {
        childIndexToKeep: number;
        elementGuidToDelete: Guid;
    } | null;
}

export interface FlowRenderInfo {
    geometry: Geometry;
    nodes: NodeRenderInfo[];
    isTerminal: boolean;
    layoutConfig: object;
    preConnector?: ConnectorRenderInfo;
}

export interface NodeRenderInfo {
    geometry: Geometry;
    label: string;
    metadata: ElementMetadata;
    config: { isSelected: boolean; isHighlighted: boolean; isSelectable: boolean; hasError: boolean };
    nextConnector?: ConnectorRenderInfo;
    flows: FlowRenderInfo[];
    faultFlow?: FlowRenderInfo;
    isNew?: boolean;
    logicConnectors: ConnectorRenderInfo[];
    conditionOptions?: Option[];
    defaultConnectorLabel?: string;
    guid: string;
    menuOpened: boolean;
    isTerminal: boolean;
    toBeDeleted: boolean;
    node?: NodeModel;
    dynamicNodeComponent?: string;
    dynamicNodeComponentSelector?: Function;
}

export interface Dimension {
    w: number;
    h: number;
}

export enum ConnectorVariant {
    DEFAULT = 'default',
    EDGE = 'edge',
    EDGE_BOTTOM = 'edgeBottom',
    CENTER = 'center',
    LOOP = 'loop',
    FAULT = 'fault',
    POST_MERGE = 'postMerge',
    POST_MERGE_TAIL = 'postMergeTail',
    BRANCH_TAIL = 'branchTail',
    BRANCH_HEAD = 'branchHead',
    BRANCH_HEAD_EMPTY = 'branchHeadEmpty',
    DEFAULT_LABEL = 'defaultLabel'
}
export enum VerticalAlign {
    TOP,
    BOTTOM
}

export interface ConnectorTypeLayoutConfig {
    h: number;
    addOffset: number;
    labelOffset?: number;
    variants?: {
        [key in ConnectorVariant]?: any;
    };
    svgMarginTop?: number;
    svgMarginBottom?: number;
}

export interface LayoutConfig {
    grid: {
        w: number;
        h: number;
    };
    menu: {
        marginBottom: number;
    };
    connector: {
        icon: {
            w: number;
            h: number;
        };
        strokeWidth: number;
        curveRadius: number;
        menu: Dimension;
        types: {
            [key in ConnectorType]?: ConnectorTypeLayoutConfig;
        };
    };
    node: {
        icon: {
            w: number;
            h: number;
        };
        w: number;
        offsetX: number;
    };
    branch: {
        defaultWidth: number;
        emptyWidth: number;
    };
}

export interface Option {
    label: string;
    value: Guid;
}

export interface MenuInfo {
    guid: NodeRef;
    menuOpened: boolean;
}

export interface ConnectorAddInfo {
    offsetY: number;
    menuOpened: boolean;
}

export interface SelectInfo {
    value: NodeRef;
    options: Option[];
}

export interface ConnectorConnectionInfo {
    prev?: NodeRef;
    next?: NodeRef;
    parent?: NodeRef;
    childIndex?: number;
}

export interface ConnectorRenderInfo {
    type: ConnectorType;
    labelType: ConnectorLabelType;
    geometry: Geometry;
    svgInfo: SvgInfo;
    addInfo?: ConnectorAddInfo;
    connectionInfo: ConnectorConnectionInfo;
    isFault: boolean;
    labelOffsetY?: number;
    connectorBadgeLabel?: string;
    isNew?: boolean;
    toBeDeleted: boolean;
    isHighlighted: boolean;
}

function getBranchLayoutKey(parentGuid: string, childIndex: number) {
    return `${parentGuid}:${childIndex}`;
}

function getLayoutByKey(key: string, progress: number, nodeLayoutMap: NodeLayoutMap): LayoutInfo {
    const { prevLayout, layout } = nodeLayoutMap[key];

    return !prevLayout || progress === 1 ? layout : tween(prevLayout, layout, progress);
}

export function tweenValue(prevValue: number, value: number, progress: number): number {
    const delta = (value - prevValue) * progress;
    return prevValue + delta;
}

/**
 * Tweens LayoutInfo values
 *
 * @param prevLayout - The previous LayoutInfo
 * @param layout - The current LayoutInfo
 * @param progress - The rendering progress
 * @returns A LayoutInfo that with the tweened values
 */
function tween(prevLayout: LayoutInfo, layout: LayoutInfo, progress: number): LayoutInfo {
    const { x, y, w, h, offsetX, joinOffsetY, addOffset, labelOffset } = layout;

    return {
        x: tweenValue(prevLayout.x, x, progress),
        y: tweenValue(prevLayout.y, y, progress),
        w: tweenValue(prevLayout.w, w, progress),
        h: tweenValue(prevLayout.h, h, progress),
        offsetX: tweenValue(prevLayout.offsetX, offsetX, progress),
        joinOffsetY: tweenValue(prevLayout.joinOffsetY, joinOffsetY, progress),
        addOffset: prevLayout.addOffset !== 0 ? tweenValue(prevLayout.addOffset, addOffset, progress) : addOffset,
        labelOffset:
            prevLayout.labelOffset !== 0 ? tweenValue(prevLayout.labelOffset, labelOffset, progress) : labelOffset
    };
}

function createDefaultLayout(): LayoutInfo {
    return {
        x: 0,
        y: 0,
        w: 0,
        h: 0,
        joinOffsetY: 0,
        offsetX: 0,
        addOffset: 0,
        labelOffset: 0
    };
}

function getLayout(nodeGuid: string, progress: number, nodeLayoutMap: NodeLayoutMap): LayoutInfo {
    if (!nodeLayoutMap[nodeGuid]) {
        return createDefaultLayout();
    }
    return getLayoutByKey(nodeGuid, progress, nodeLayoutMap);
}

/**
 * Get a LayoutInfo for a branch
 *
 * @param parentGuid - A branching element guid
 * @param childIndex  - A branch index
 * @param progress - Rendering progress
 * @param nodeLayoutMap - The node layout map
 * @returns A LayoutInfo for the branch
 */
function getBranchLayout(
    parentGuid: Guid,
    childIndex: number,
    progress: number,
    nodeLayoutMap: NodeLayoutMap
): LayoutInfo {
    const key = getBranchLayoutKey(parentGuid, childIndex);
    if (!nodeLayoutMap[key]) {
        return createDefaultLayout();
    }
    return getLayoutByKey(key, progress, nodeLayoutMap);
}

/**
 * Get the config for a connector type and variants
 *
 * @param layoutConfig - The layout config
 * @param connectorType - The connector type
 * @param connectorVariants - A list of variants
 * @return A connector layout config for the connector
 */
function getConnectorConfig(
    layoutConfig: LayoutConfig,
    connectorType: ConnectorType,
    ...connectorVariants: ConnectorVariant[]
): ConnectorTypeLayoutConfig {
    let connectorConfig = layoutConfig.connector.types[connectorType]!;

    // go through variants and override values
    connectorVariants.forEach((connectorVariant: ConnectorVariant) => {
        const variantConfig = connectorConfig.variants![connectorVariant];
        if (variantConfig != null) {
            connectorConfig = { ...connectorConfig, ...variantConfig } as ConnectorTypeLayoutConfig;
        }
    });

    return connectorConfig;
}

/**
 * For a branching element, get the number of outcomes that merge back
 *
 * @param flowModel - The flow model
 * @param branchingElement - THe branching element
 * @return the number of outcomes that merge back
 */
const getMergeOutcomeCount = (flowModel: FlowModel, branchingElement: ParentNodeModel) =>
    branchingElement.children.reduce((count, child) => {
        return child == null || !(flowModel[child] as BranchHeadNodeModel).isTerminal ? count + 1 : count;
    }, 0);

export {
    getBranchLayoutKey,
    getBranchLayout,
    tween,
    createDefaultLayout,
    getLayout,
    getConnectorConfig,
    getMergeOutcomeCount
};
