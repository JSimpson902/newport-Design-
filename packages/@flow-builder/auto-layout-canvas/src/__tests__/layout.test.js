import { getBranchLayoutKey } from '../flowRendererUtils';
import { calculateFlowLayout, getLayoutChildOrFault } from '../layout';
import { FAULT_INDEX } from '../model';
import {
    BRANCH_ELEMENT,
    createFlow,
    createFlowRenderContext,
    END_ELEMENT,
    getEmptyFlowContext,
    getFlowWhenGoingFromParentDefaultBranchToPreviousElement,
    getFlowWhenGoingFromParentFaultBranchToPreviousElement,
    getFlowWhenGoingFromParentFirstBranchToPreviousElement,
    getFlowWhenGoingToPreviousElement,
    getFlowWithDecisionWithOneElementOnLeftBranchContext,
    getFlowWithDynamicNodeComponent,
    getFlowWithEmptyDeciisionWith3BranchesContext,
    getFlowWithEmptyDecisionContext,
    getFlowWithEmptyLoopContext,
    getFlowWithOnlyImmediateScheduledPathContext,
    getFlowWithScheduledPathsContext,
    getFlowWithTwoFaults,
    LOOP_ELEMENT
} from './testUtils';

function calculateLayoutAndAssert(ctx) {
    calculateFlowLayout(ctx);
    expect(ctx.nodeLayoutMap).toMatchSnapshot();
}

describe('layout', () => {
    describe('getBranchLayoutKey', () => {
        it('creates a key', () => {
            expect(getBranchLayoutKey('guid', 2)).toEqual('guid:2');
        });
    });

    describe('getLayoutChildOrFault', () => {
        const parentNodeModel = {
            children: ['child-guid0', 'child-guid1'],
            fault: 'fault-guid'
        };

        it('returns a child guid when non-negative', () => {
            expect(getLayoutChildOrFault(parentNodeModel, 1)).toEqual('child-guid1');
        });

        it('returns the fault when -1', () => {
            expect(getLayoutChildOrFault(parentNodeModel, FAULT_INDEX)).toEqual('fault-guid');
        });
    });

    describe('calculateLayout', () => {
        it('empty flow', () => {
            calculateLayoutAndAssert(getEmptyFlowContext());
        });

        it('flow with empty decision', () => {
            calculateLayoutAndAssert(getFlowWithEmptyDecisionContext());
        });

        it('flow with empty loop', () => {
            calculateLayoutAndAssert(getFlowWithEmptyLoopContext());
        });

        it('flow with loop and nested loop', () => {
            const nestedLoopElement = { ...LOOP_ELEMENT, children: [null] };
            const loopElement = { ...LOOP_ELEMENT, children: [[nestedLoopElement]] };

            const flowModel = createFlow([loopElement]);
            calculateLayoutAndAssert(createFlowRenderContext({ flowModel }));
        });

        it('flow with loop and nested loop - loop menu opened', () => {
            const nestedLoopElement = { ...LOOP_ELEMENT, children: [null] };
            const loopElement = { ...LOOP_ELEMENT, children: [[nestedLoopElement]] };

            const flowModel = createFlow([loopElement]);
            const interactionState = {
                menuInfo: {
                    key: 'loop-guid',
                    type: 0,
                    geometry: {
                        w: 300,
                        h: 221,
                        x: 0,
                        y: 0
                    }
                }
            };
            calculateLayoutAndAssert(createFlowRenderContext({ flowModel, interactionState }));
        });

        it('flow with loop with ended branch', () => {
            const loopElement = { ...LOOP_ELEMENT, children: [[END_ELEMENT]] };

            const flowModel = createFlow([loopElement]);
            calculateLayoutAndAssert(createFlowRenderContext({ flowModel }));
        });

        it('flow with decision and ended middle branch', () => {
            const branchElement = { ...BRANCH_ELEMENT, children: [null, [END_ELEMENT], null] };

            const flowModel = createFlow([branchElement]);
            calculateLayoutAndAssert(createFlowRenderContext({ flowModel }));
        });

        it('flow with empty decision with 3 branches', () => {
            calculateLayoutAndAssert(getFlowWithEmptyDeciisionWith3BranchesContext());
        });

        it('flow with decision with one element on left branch', () => {
            calculateLayoutAndAssert(getFlowWithDecisionWithOneElementOnLeftBranchContext());
        });

        it('flow with two elements with fault branches', () => {
            calculateLayoutAndAssert(getFlowWithTwoFaults());
        });

        it('flow with decision with 3 branches where the middle branch is a left bottom edge', () => {
            const nestedBranch = { ...BRANCH_ELEMENT, children: [null, null, null] };
            const branchElement = { ...BRANCH_ELEMENT, children: [[END_ELEMENT], null, [nestedBranch]] };

            const flowModel = createFlow([branchElement]);
            calculateLayoutAndAssert(createFlowRenderContext({ flowModel }));
        });

        it('flow with decision with 3 branches with decision menu opened', () => {
            const nestedBranch = { ...BRANCH_ELEMENT, children: [null, null, null] };
            const branchElement = { ...BRANCH_ELEMENT, children: [[END_ELEMENT], null, [nestedBranch]] };

            const flowModel = createFlow([branchElement]);
            const interactionState = {
                menuInfo: {
                    key: 'branch-guid',
                    type: 0,

                    geometry: {
                        w: 300,
                        h: 221,
                        x: 0,
                        y: 0
                    }
                }
            };
            calculateLayoutAndAssert(createFlowRenderContext({ flowModel, interactionState }));
        });

        it('flow with dynamic node component', () => {
            calculateLayoutAndAssert(getFlowWithDynamicNodeComponent());
        });

        it('flow with a GoTo connection to the previous element', () => {
            calculateLayoutAndAssert(getFlowWhenGoingToPreviousElement(false, false));
        });

        it('flow with a GoTo connection to the previous element and source node menu is open', () => {
            calculateLayoutAndAssert(getFlowWhenGoingToPreviousElement(true, false));
        });

        it('flow with a GoTo connection to the previous element and goTo connector menu is open', () => {
            calculateLayoutAndAssert(getFlowWhenGoingToPreviousElement(false, true));
        });

        it('flow with a GoTo connection from parent first branch to the previous element', () => {
            calculateLayoutAndAssert(getFlowWhenGoingFromParentFirstBranchToPreviousElement(false, false, false));
        });

        it('flow with a GoTo connection from parent first branch to the previous element and the source node menu is open', () => {
            calculateLayoutAndAssert(getFlowWhenGoingFromParentFirstBranchToPreviousElement(true, false, false));
        });

        it('flow with a GoTo connection from parent first branch to the previous element when merge connector menu is open', () => {
            calculateLayoutAndAssert(getFlowWhenGoingFromParentFirstBranchToPreviousElement(false, true, false));
        });

        it('flow with a GoTo connection from parent first branch to the previous element when branch head connector menu is open', () => {
            calculateLayoutAndAssert(getFlowWhenGoingFromParentFirstBranchToPreviousElement(false, false, true));
        });

        it('flow with a GoTo connection from parent default branch to the previous element', () => {
            calculateLayoutAndAssert(getFlowWhenGoingFromParentDefaultBranchToPreviousElement());
        });

        it('flow with a GoTo connection from parent fault branch to the previous element', () => {
            calculateLayoutAndAssert(getFlowWhenGoingFromParentFaultBranchToPreviousElement());
        });

        it('flow with scheduled paths', () => {
            calculateLayoutAndAssert(getFlowWithScheduledPathsContext());
        });

        it('flow with only the immediate scheduled path with start menu closed', () => {
            calculateLayoutAndAssert(getFlowWithOnlyImmediateScheduledPathContext());
        });

        it('flow with only the immediate scheduled path with start menu opened', () => {
            calculateLayoutAndAssert(getFlowWithOnlyImmediateScheduledPathContext(true));
        });
    });
});
