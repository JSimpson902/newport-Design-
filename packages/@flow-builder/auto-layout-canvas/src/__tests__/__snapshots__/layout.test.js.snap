// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout calculateLayout empty flow 1`] = `
Object {
  "end-guid": Object {
    "layout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 120,
    },
    "prevLayout": undefined,
  },
  "root:0": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 0,
    },
  },
  "start-guid": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 0,
    },
    "prevLayout": undefined,
  },
}
`;

exports[`layout calculateLayout flow with decision with one element on left branch 1`] = `
Object {
  "branch-guid": Object {
    "layout": Object {
      "h": 336,
      "joinOffsetY": 240,
      "offsetX": 220,
      "w": 440,
      "x": 0,
      "y": 120,
    },
    "prevLayout": undefined,
  },
  "branch-guid:0": Object {
    "layout": Object {
      "h": 240,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": -132,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "branch-guid:1": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": 132,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "branch-left-head-guid": Object {
    "layout": Object {
      "h": 96,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 144,
    },
    "prevLayout": undefined,
  },
  "end-guid": Object {
    "layout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 456,
    },
    "prevLayout": undefined,
  },
  "root:0": Object {
    "layout": Object {
      "h": 456,
      "joinOffsetY": 0,
      "offsetX": 220,
      "w": 440,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 0,
    },
  },
  "start-guid": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 0,
    },
    "prevLayout": undefined,
  },
}
`;

exports[`layout calculateLayout flow with empty decision 1`] = `
Object {
  "branch-guid": Object {
    "layout": Object {
      "h": 216,
      "joinOffsetY": 120,
      "offsetX": 176,
      "w": 352,
      "x": 0,
      "y": 120,
    },
    "prevLayout": undefined,
  },
  "branch-guid:0": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": -88,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "branch-guid:1": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": 88,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "end-guid": Object {
    "layout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 336,
    },
    "prevLayout": undefined,
  },
  "root:0": Object {
    "layout": Object {
      "h": 336,
      "joinOffsetY": 0,
      "offsetX": 176,
      "w": 352,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 0,
    },
  },
  "start-guid": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 0,
    },
    "prevLayout": undefined,
  },
}
`;

exports[`layout calculateLayout flow with empty decision with 3 branches 1`] = `
Object {
  "branch-guid": Object {
    "layout": Object {
      "h": 216,
      "joinOffsetY": 120,
      "offsetX": 264,
      "w": 528,
      "x": 0,
      "y": 120,
    },
    "prevLayout": undefined,
  },
  "branch-guid:0": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": -176,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "branch-guid:1": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "branch-guid:2": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": 176,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "end-guid": Object {
    "layout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 336,
    },
    "prevLayout": undefined,
  },
  "root:0": Object {
    "layout": Object {
      "h": 336,
      "joinOffsetY": 0,
      "offsetX": 264,
      "w": 528,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 0,
    },
  },
  "start-guid": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 0,
    },
    "prevLayout": undefined,
  },
}
`;

exports[`layout calculateLayout flow with empty loop 1`] = `
Object {
  "end-guid": Object {
    "layout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 360,
    },
    "prevLayout": undefined,
  },
  "loop-guid": Object {
    "layout": Object {
      "h": 240,
      "joinOffsetY": 144,
      "offsetX": 148,
      "w": 296,
      "x": 0,
      "y": 120,
    },
    "prevLayout": undefined,
  },
  "loop-guid:0": Object {
    "layout": Object {
      "h": 144,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 176,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 120,
    },
  },
  "root:0": Object {
    "layout": Object {
      "h": 360,
      "joinOffsetY": 0,
      "offsetX": 148,
      "w": 324,
      "x": 0,
      "y": 0,
    },
    "prevLayout": Object {
      "h": 0,
      "joinOffsetY": 0,
      "offsetX": 0,
      "w": 0,
      "x": 0,
      "y": 0,
    },
  },
  "start-guid": Object {
    "layout": Object {
      "h": 120,
      "joinOffsetY": 0,
      "offsetX": 88,
      "w": 264,
      "x": 0,
      "y": 0,
    },
    "prevLayout": undefined,
  },
}
`;
